### PR TITLE
Add `Arrayable` type #270

### LINKED ISSUE #270: Like `Promisable<X>` but for arrays: `X | X[]`
I remember seeing this type somewhere, but I can't find it. Does it exist? Is it native?

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/270">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/270/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Max10240**
fix #270

### CODE REVIEW

File: source/arrayable.d.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/**
+Create a type that represents either the value or array of the value.
```
Comment (sindresorhus) [inline source/arrayable.d.ts:2]:
```suggestion
Create a type that represents either the value or an array of the value.
```

File: source/arrayable.d.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+@see Promisable
+
+@example
+```
+import type {Arrayable} from 'type-fest';
+
+function bundle(input: string, output: Arrayable<string>) {
+  const outputList = Array.isArray(output) ? output : [output];
```
Comment (sindresorhus) [inline source/arrayable.d.ts:11]: Tab-indentation

File: source/arrayable.d.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  }
+}
+
+bundle('src/index.js', 'dist/index.js');
+bundle('src/index.js', ['dist/index.cjs', 'dist/index.mjs']);
+```
+
+@category Array
+*/
+export type Arrayable<T> = T | T[];
```
Comment (sindresorhus) [inline source/arrayable.d.ts:25]:
```suggestion
export type Arrayable<T> = T | readonly T[];
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Review by Emiyaaaaa — APPROVED** [review]

**Comment by fregante** [comment]
@sindresorhus why was `readonly []` chosen here? I'm unable to create a `castArray` function that is compatible with this.

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwIJSgQwJ5oEYBsCmAvnAGZQQhwDkMGY+AtCfgM4xUDcAUFyQK4A7AMYxgEAXCFo2KdBgCMAHgCSMfCAB8ACgBuaXH3wAuOLMw4CKtZoCUJ1eoDaAXURcAkFHww+UCWYwAOmAWAN19Qxs4AH44PQN8OBNHeMNnbkIefmFRcUlpGACAJit1bVTjU1RzPHxS2xMvNAATcVwMOAcQFzdPb19-aqCQsIqo2IqkuBSI-HSuTKA

**Comment by Max10240** [comment]
@fregante  `Item` cannot be an array type, because `Arrayable<unknown[]>` is equivalent to `unknown[] | readonly unknown[][]`, NOT `unknown[]` or `readonly unknown[]`. If `Item` is an Array type, `Array.isArray` will always return `true`, meaning that the actual return type of `castArray1 / castArray2` is actually `Item` itself, not `Item[]`.
I'm guessing what you're looking for is something like:
```tsx
import { Arrayable } from 'type-fest';

type Primitive =
	| bigint
	| boolean
	| null
	| number
	| string
	| symbol
	| undefined;

function castArray<Item extends Primitive | Record<PropertyKey, unknown>>(value: Arrayable<Item>): Item[] {
	return Array.isArray(value) ? value : [value];
}

```

**Comment by fregante** [comment]
My previous code worked correctly if `readonly` is dropped, so that lengthy type information is a consequence of this keyword. 

You can see it working here: https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbwIJSgQwJ5oEYBsCmAvnAGZQQhwDkMGY+AtCfgM4xUDcAUF7fXAAVcaYADsU6LHnwAeABoA+OAF44cuAB81AbQC63HiQCuogMYxgEUXFNo2EzAEYZASRj4QCgBQA3NLiN8AC5BYTEHKQJXd08AShC3Dz1ELgBIKHwYIyhrCIA6YBYI339A2LgAfjg-APw4EO0awP0uQkMTc0trW3tUTAAmaI9vJuC4CJwoxLiEmOSENIysnPG+jAKitZLa8qrR+rhG0vwWtqA

**Pull request merged**